### PR TITLE
Modifications to use the new field split capability of ESMF #2454

### PR DESCRIPTION
**Pull request by aoloso**
Changes to how fields are allocated to use farray instead of farrayPtr
## Description

Made changes to allocate fields to use farray instead of farrayPtr. This allows explicit specification of indexflag required by the new MAPL field split functionality. This functionality allows a clean way to create a new field from an exiting field where the new field is a 'slice' of the existing field with the slicing index being that of the trailing ungiridded dim of the existing field.
## Motivation and Context

Provides a clean, succinct, general, and less error prone way to create a new field from an existing field.
## How Has This Been Tested?

zero-diff with the develop branch through c720

Skylake nodes on discover SLES 12 with stock g5_modules but Baselibs-7.16.0, i.e.
      comp/gcc/11.2.0
      comp/intel/2021.6.0
      mpi/impi/2021.6.0
      python/GEOSpyD/Min4.11.0_py3.9_AND_Min4.8.3_py2.7
      /discover/swdev/gmao_SIteam/Baselibs/ESMA-Baselibs-7.16.0/x86_64-pc-linux-gnu/ifort_2021.6.0-intelmpi_2021.6.0-SLES12

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

**Comment by mathomp4** [comment]
Okay. This means I need to get an ESMA_env up and ready with ESMF 8.6.0. I guess this is a good place to test it...

File: base/Base/Base_Base_implementation.F90 [hunk 1]
```diff
@@ -192,9 +192,12 @@ module subroutine MAPL_FieldAllocCommit(field, dims, location, typekind, &
...
              _VERIFY(STATUS)
              VAR_1D = INIT_VALUE
-             call ESMF_FieldEmptyComplete(FIELD, farrayPtr=VAR_1D,    &
+             call ESMF_FieldEmptyComplete(FIELD, farray=VAR_1D,    &
+                  indexflag=ESMF_INDEX_DELOCAL, &
                   datacopyFlag = ESMF_DATACOPY_REFERENCE,             &
                   gridToFieldMap=gridToFieldMap,                      &
+                  ungriddedLBound=[1],&
+                  ungriddedUBound=[ungrid(1)], &
                   rc = status)
```
Comment (tclune) [inline base/Base/Base_Base_implementation.F90:201]: It looks like `status` here is not being checked.  As long as we are here, please fix this.

File: base/Base/Base_Base_implementation.F90 [hunk 2]
```diff
@@ -206,9 +209,12 @@ module subroutine MAPL_FieldAllocCommit(field, dims, location, typekind, &
...
              _VERIFY(STATUS)
              VR8_1D = INIT_VALUE
-             call ESMF_FieldEmptyComplete(FIELD, farrayPtr=VR8_1D,    &
+             call ESMF_FieldEmptyComplete(FIELD, farray=VR8_1D,    &
+                  indexflag=ESMF_INDEX_DELOCAL, &
                   datacopyFlag = ESMF_DATACOPY_REFERENCE,             &
                   gridToFieldMap=gridToFieldMap,                      &
+                  ungriddedLBound=[1],&
+                  ungriddedUBound=[ungrid(1)], &
                   rc = status)
```
Comment (tclune) [inline base/Base/Base_Base_implementation.F90:218]: Again - please check status.

File: base/Base/Base_Base_implementation.F90 [hunk 3]
```diff
@@ -242,17 +248,25 @@ module subroutine MAPL_FieldAllocCommit(field, dims, location, typekind, &
           _VERIFY(STATUS)
           VAR_1D = INIT_VALUE
 
-          call ESMF_FieldEmptyComplete(FIELD, farrayPtr=var_1d,  &
+          call ESMF_FieldEmptyComplete(FIELD, farray=var_1d,  &
+               indexflag=ESMF_INDEX_DELOCAL, &
+               datacopyFlag = ESMF_DATACOPY_REFERENCE,             &
                gridToFieldMap=gridToFieldMap,                           &
+               ungriddedLBound=[lb1],&
+               ungriddedUBound=[ub1], &
```
Comment (tclune) [inline base/Base/Base_Base_implementation.F90:256]: Prefer `_RC` to `_VERIFY` as long as you are making changes anyway.

File: base/Base/Base_Base_implementation.F90 [hunk 4]
```diff
@@ -242,17 +248,25 @@ module subroutine MAPL_FieldAllocCommit(field, dims, location, typekind, &
...
           VR8_1D = INIT_VALUE
 
-          call ESMF_FieldEmptyComplete(FIELD, farrayPtr=vr8_1d,  &
+          call ESMF_FieldEmptyComplete(FIELD, farray=vr8_1d,  &
+               indexflag=ESMF_INDEX_DELOCAL, &
+               datacopyFlag = ESMF_DATACOPY_REFERENCE,             &
                gridToFieldMap=gridToFieldMap,                           &
+               ungriddedLBound=[lb1],&
+               ungriddedUBound=[ub1], &
                RC=status)
```
Comment (tclune) [inline base/Base/Base_Base_implementation.F90:270]: Check status.

File: base/Base/Base_Base_implementation.F90 [hunk 5]
```diff
@@ -287,21 +302,27 @@ module subroutine MAPL_FieldAllocCommit(field, dims, location, typekind, &
...
-             call ESMF_FieldEmptyComplete(FIELD, farrayPtr=VAR_3D, &
+             call ESMF_FieldEmptyComplete(FIELD, farray=VAR_3D, &
+                  indexflag=ESMF_INDEX_DELOCAL, &
                   datacopyFlag = ESMF_DATACOPY_REFERENCE,         &
                   gridToFieldMap=gridToFieldMap,              &
                   totalLWidth=haloWidth(1:griddedDims),     &
                   totalUWidth=haloWidth(1:griddedDims),     &
+                  ungriddedLBound=[1],&
+                  ungriddedUBound=[ungrid(1)], &
                   rc = status)
```
Comment (tclune) [inline base/Base/Base_Base_implementation.F90:313]: Looks like there are a bunch of these.  Please do a sweep of the file and fix.  (Feel free to make this status stuff a separate PR - just let me know.)

**Review by tclune — CHANGES_REQUESTED** [review]
It will be nice when templates let us avoid so much duplication.

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed - 2023-11-28
+Made changes to allocate fields to use farray instead of farrayPtr. This allows explicit specification of indexflag required by the new MAPL field split functionality. This functionality allows a clean way to create a new field from an exiting field where the new field is a 'slice' of the existing field with the slicing index being that of the trailing ungiridded dim of the existing field.
```
[the remark this answers is not in the document]
Comment (mathomp4) [reply CHANGELOG.md:11]: Also, no need for a date-stamp in that. Just gets under `### Changed` like everything else

[the remark this answers is not in the document]
Comment (tclune) [reply CHANGELOG.md:11]: @aoloso You are closer to it.  I just wanted to make certain you had considered.   I thought you had introduced a new option (optional argument) to some interface which sounds a bit more like "Added".  But if there are no changes to the interfaces than "Changed" is right.

File: base/Base/Base_Base_implementation.F90 (hunk 1 above)
Comment (aoloso) [reply base/Base/Base_Base_implementation.F90:201]: Satisfied in 224 at the end of the 'if' block

File: base/Base/Base_Base_implementation.F90 (hunk 2 above)
Comment (aoloso) [reply base/Base/Base_Base_implementation.F90:218]: Satisfied in 224 at the end of the 'if' block

Comment (tclune) [reply base/Base/Base_Base_implementation.F90:218]: Yes, but ... this is somewhat confusing to the reader.  They then have to go back and see how many different places that failure could have come from.  Best to do it right away (in source) unless there is a pressing reason to not do so.

File: base/Base/Base_Base_implementation.F90 (hunk 5 above)
Comment (aoloso) [reply base/Base/Base_Base_implementation.F90:313]: I have cleaned the file up to replace all RC/_VERIFY with _RC and all STAT/_VERIFY with _STAT. I will mark them as resolved here and make a new PR just for that.

**Review by tclune — APPROVED** [review]

**Comment by tclune** [comment]
@aoloso Now you need to dig into why the CI tests fail ...

**Review by tclune — APPROVED** [review]

**Review by tclune — DISMISSED** [review]

**Pull request closed without merging**